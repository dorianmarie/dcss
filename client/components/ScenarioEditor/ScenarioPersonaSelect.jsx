import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Dropdown, Table } from '@components/UI';
import * as Color from '@utils/Color';

const ScenarioPersonaSelectItem = ({
  name,
  description,
  color: backgroundColor
}) => {
  const color = Color.foregroundColor(backgroundColor);

  return (
    <Table>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell style={{ color, backgroundColor }}>
            {name}
          </Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {description ? (
          <Table.Row>
            <Table.Cell>{description}</Table.Cell>
          </Table.Row>
        ) : null}
      </Table.Body>
    </Table>
  );
};

ScenarioPersonaSelectItem.propTypes = {
  color: PropTypes.string,
  description: PropTypes.string,
  name: PropTypes.string
};

class ScenarioPersonaSelect extends Component {
  constructor(props) {
    super(props);

    this.personasById = props.personas.reduce((accum, persona) => {
      accum[persona.id] = persona;
      return accum;
    }, {});

    this.onChange = this.onChange.bind(this);
  }

  onChange(e, { value }) {
    this.props.onSelect((value && this.personasById[value]) || null);
  }

  render() {
    const { onChange } = this;
    const {
      defaultValue,
      emptyText = 'No selection',
      fluid,
      item,
      personas,
      placeholder,
      search,
      selection,
      value
    } = this.props;

    // This empty "persona" option is used to prevent the
    // dropdown from treating "open" as "select".
    personas.unshift({
      id: '',
      value: null,
      name: emptyText
    });

    const options = personas.reduce((accum, persona) => {
      const option = {
        key: persona.id,
        value: persona.id,
        text: persona.name
      };

      if (persona.id && persona.name) {
        option.content = <ScenarioPersonaSelectItem {...persona} />;
      }
      return accum.concat([option]);
    }, []);

    const scrolling = true;
    const dropdownProps = {
      'aria-label': placeholder,
      defaultValue,
      fluid,
      item,
      onChange,
      options,
      placeholder,
      scrolling,
      search,
      selection,
      value
    };

    return <Dropdown className="sp__dropdown" {...dropdownProps} />;
  }
}

ScenarioPersonaSelect.propTypes = {
  defaultValue: PropTypes.node,
  emptyText: PropTypes.string,
  fluid: PropTypes.bool,
  item: PropTypes.any,
  onSelect: PropTypes.func,
  personas: PropTypes.array,
  placeholder: PropTypes.node,
  scenario: PropTypes.object,
  search: PropTypes.bool,
  selection: PropTypes.bool,
  style: PropTypes.object,
  value: PropTypes.node
};

const mapStateToProps = (state, ownProps) => {
  const { scenario } = state;
  const personas = ownProps.personas.slice(0);
  return {
    personas,
    scenario
  };
};

export default connect(
  mapStateToProps,
  null
)(ScenarioPersonaSelect);
