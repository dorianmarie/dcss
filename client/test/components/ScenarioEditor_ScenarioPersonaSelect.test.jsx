import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/** @GENERATED: BEGIN **/

const personas = [
  {
    id: 1,
    name: 'Participant',
    description: 'The default user participating in a single person scenario.',
    color: '#FFFFFF',
    created_at: '2020-12-01T15:49:04.962Z',
    updated_at: null,
    deleted_at: null,
    author_id: 3,
    is_read_only: true,
    is_shared: true
  },
  {
    id: 2,
    name: 'Teacher',
    description:
      'A non-specific teacher, participating in a multi person scenario.',
    color: '#3f59a9',
    created_at: '2020-12-01T15:49:04.962Z',
    updated_at: null,
    deleted_at: null,
    author_id: 3,
    is_read_only: true,
    is_shared: true
  },
  {
    id: 3,
    name: 'Student',
    description:
      'A non-specific student, participating in a multi person scenario.',
    color: '#e59235',
    created_at: '2020-12-01T15:49:04.962Z',
    updated_at: null,
    deleted_at: null,
    author_id: 3,
    is_read_only: true,
    is_shared: true
  }
];

import ScenarioPersonaSelect from '../../components/ScenarioEditor/ScenarioPersonaSelect.jsx';
/** @GENERATED: END **/

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  /** @GENERATED: BEGIN **/

  /** @GENERATED: END **/

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('ScenarioPersonaSelect', () => {
  expect(ScenarioPersonaSelect).toBeDefined();
});

test('Render 1 1', async done => {
  /** @GENERATED: BEGIN **/
  const Component = ScenarioPersonaSelect;
  const props = {
    ...commonProps,
    personas,
    onSelect: jest.fn()
  };

  const state = {
    ...commonState
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);
  /** @GENERATED: END **/

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();

  done();
});
