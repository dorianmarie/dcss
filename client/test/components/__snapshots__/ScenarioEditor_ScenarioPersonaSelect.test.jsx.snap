// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <div
      aria-expanded=\\"false\\"
      class=\\"ui scrolling dropdown sp__dropdown\\"
      role=\\"listbox\\"
      tabindex=\\"0\\"
    >
      <i
        aria-hidden=\\"true\\"
        class=\\"dropdown icon\\"
      />
      <div
        class=\\"menu transition\\"
      >
        <div
          aria-checked=\\"true\\"
          aria-selected=\\"true\\"
          class=\\"active selected item\\"
          role=\\"option\\"
          style=\\"pointer-events: all;\\"
        >
          <span
            class=\\"text\\"
          >
            No selection
          </span>
        </div>
        <div
          aria-checked=\\"false\\"
          aria-selected=\\"false\\"
          class=\\"item\\"
          role=\\"option\\"
          style=\\"pointer-events: all;\\"
        >
          <table
            class=\\"ui table\\"
          >
            <thead
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <th
                  class=\\"\\"
                  style=\\"color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);\\"
                >
                  Participant
                </th>
              </tr>
            </thead>
            <tbody
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <td
                  class=\\"\\"
                >
                  The default user participating in a single person scenario.
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          aria-checked=\\"false\\"
          aria-selected=\\"false\\"
          class=\\"item\\"
          role=\\"option\\"
          style=\\"pointer-events: all;\\"
        >
          <table
            class=\\"ui table\\"
          >
            <thead
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <th
                  class=\\"\\"
                  style=\\"color: rgb(255, 255, 255); background-color: rgb(63, 89, 169);\\"
                >
                  Teacher
                </th>
              </tr>
            </thead>
            <tbody
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <td
                  class=\\"\\"
                >
                  A non-specific teacher, participating in a multi person scenario.
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          aria-checked=\\"false\\"
          aria-selected=\\"false\\"
          class=\\"item\\"
          role=\\"option\\"
          style=\\"pointer-events: all;\\"
        >
          <table
            class=\\"ui table\\"
          >
            <thead
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <th
                  class=\\"\\"
                  style=\\"color: rgb(0, 0, 0); background-color: rgb(229, 146, 53);\\"
                >
                  Student
                </th>
              </tr>
            </thead>
            <tbody
              class=\\"\\"
            >
              <tr
                class=\\"\\"
              >
                <td
                  class=\\"\\"
                >
                  A non-specific student, participating in a multi person scenario.
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</body>"
`;
